doc/developer/design/20230227_external_introspection.md

```diff
@@ -0,0 +1,175 @@
+
+# Reference explanation
+[reference-explanation]: #reference-explanation
+
+## Data sources
+All features will be implemented in the web UI using React, querying
+the `mz_internal` relations for the necessary data. Relations that
+will be used include:
+
+* For the hierarchical dataflow visualizer:
+    * `mz_dataflow_operators`
+    * `mz_dataflow_channels`
+    * `mz_arrangement_sizes`
+    * `mz_dataflow_addresses`
+* For the global frontier lag visualizer:
+    * `mz_object_dependencies`
+    * `mz_cluster_replica_frontiers`
+* For the flamegraph views: same as the hierarchical dataflow
+  visualizer
+* For the per-query metadata: New table to be created
+  (`mz_query_metadata`).
+* For the user-friendly plan rendering: parsed `EXPLAIN PLAN` output
+  (possibly from `mz_query_metadata`, or entered manually by the user).
+
+## Layout and rendering
+
+We will use the [d3-graphviz](https://github.com/magjac/d3-graphviz)
+library for layout and rendering of the hierarchical dataflow
+visualizer, global frontier lag visualizer, and user-friendly plan
+rendering. We will use
+[d3-flame-graph](https://github.com/spiermar/d3-flame-graph) -- which
+we are already using on the internal side -- for the flamegraph
+visualizer.
+
+# Rollout and Lifecycle
+
+We will be considering this an experimental feature until such time as
+we can involve professional product designers and front-end
+engineers. Until that time, the level of polish of the UX may not
+reach the same standards as the rest of the site, so we will require
+users to click a link with a label like "Advanced Features" or similar
+in order to access the tools.
+
+We will also use LaunchDarkly to gate access to the feature, and not
+launch it at all to the public until we have gotten some internal
+feedback from support and DevEx that it is useful.
+
+We will have a separate LaunchDarkly flag just for the
+`mz_query_metadata` table, since this will have potentially large cost
+in high-QPS scenarios.
+
+## Testing and observability
+
+(TBD -- will cover this in our meeting with Robin tomorrow and get an
+overview from him of the testing strategy for this kind of feature)
+
+# Drawbacks
+[drawbacks]: #drawbacks
+
+* If any of the features don't prove useful, we are cluttering our UX
+  unnecessarily.
+* Maintaining the `mz_query_metadata` table will introduce overhead on
+  persist, especially in high-QPS scenarios. We should measure this
+  before implementing.
+
+# Conclusion and alternatives
+[conclusion-and-alternatives]: #conclusion-and-alternatives
+
+- I am unaware of any other possible designs
+
+# Unresolved questions
+[unresolved-questions]: #unresolved-questions
+
+* What will be the overhead of the `mz_query_metadata` table, and is
+  it acceptable?
+* How can we communicate query IDs back to the user for looking up the
+  query in the per-query metadata view? Is it acceptable to use
+  `NOTICE` messages in this case?
```

Comment:
It `mz_query_metadata` contains the query SQL and can be filtered by user and ordered by time, maybe that's sufficient for users to find their queries already?

Reply:
It is probably sufficient if you filter by user.

However, now I'm wondering if there is a privacy concern. If we store query metadata for all users in the same table, then anyone can query it and see what other users are doing. To give a Unix analogy, it's a bit like if your shell `$HISTFILE` was user-readable. We should check with Graham about whether this is acceptable.